File: 20-valid-parentheses/valid-parentheses.java
class Solution {

    private final static Map<Character, Character> mappings = new HashMap<>();

    static {
        mappings.put(')', '(');
        mappings.put('}', '{');
        mappings.put(']', '[');     
    }

    public boolean isValid(String s) {
        Deque<Character> st = new ArrayDeque<>();

        for (char ch : s.toCharArray()) {
            if (!mappings.containsKey(ch)) {
                st.push(ch);
            } else {
                if (st.isEmpty() || mappings.get(ch) != st.peek()) return false;
                
                st.pop();
            }
        }

        return st.isEmpty();
    }
}

File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> map = {
            {')', '('},
            {'}', '{'},
            {']', '['}
        };

        stack<char> st;

        for (char ch : s) {
            if (map.find(ch) == map.end()) {
                st.push(ch);
            } else {
                if (st.empty() || map[ch] != st.top()) {
                    return false;
                }
                st.pop();
            }
        }

        return st.empty();
    }
};